platforms/ide/tooling-api/src/main/java/org/gradle/tooling/internal/adapter/ProtocolToModelAdapter.java

```diff
@@ -159,14 +164,66 @@ private static <T> T createView(Class<T> targetType, @Nullable Object sourceObje
+        Class<?>[] modelContractInterfaces = getModelContractInterfaces(targetType, sourceObject, viewType);
+        Object proxy = Proxy.newProxyInstance(viewType.getClassLoader(), modelContractInterfaces, handler);
         handler.attachProxy(proxy);
 
         graphDetails.putViewFor(sourceObject, viewKey, proxy);
 
         return viewType.cast(proxy);
     }
 
+    private static <T> Class<?>[] getModelContractInterfaces(Class<T> targetType, Object sourceObject, Class<? extends T> viewType) {
+        Map<String, Class<?>> potentialSubInterfaces = getPotentialModelContractSubInterfaces(targetType);
+        Set<Class<?>> actualSubInterfaces = getActualImplementedModelContractSubInterfaces(sourceObject, potentialSubInterfaces);
+
+        List<Class<?>> modelContractInterfaces = new ArrayList<>();
+        modelContractInterfaces.add(viewType); // base interface
+        modelContractInterfaces.addAll(actualSubInterfaces);
+        return modelContractInterfaces.toArray(new Class<?>[0]);
+    }
+
+    private static <T> Map<String, Class<?>> getPotentialModelContractSubInterfaces(Class<T> targetType) {
+        Annotation[] annotations = targetType.getAnnotations();
+        Map<String, Class<?>> result = new HashMap<>();
+        for (Annotation annotation : annotations) {
+            if (annotation instanceof ToolingModelContract) {
+                Class<?>[] classes = ((ToolingModelContract) annotation).subTypes();
+                for (Class<?> clazz : classes) {
+                    result.put(clazz.getName(), clazz);
+                }
+            }
+        }
+        return result;
+    }
+
+    private static Set<Class<?>> getActualImplementedModelContractSubInterfaces(Object sourceObject, Map<String, Class<?>> potentialModelContractInterfaces) {
+        // deep-traverse the source object's type hierarchy and extract all implemented interfaces
+        Iterable<Class<?>> allImplementedInterfaces = Traverser.forGraph(new SuccessorsFunction<Class<?>>() {
+            @Override
+            public Iterable<? extends Class<?>> successors(Class<?> node) {
+                ImmutableList.Builder<Class<?>> builder = new ImmutableList.Builder<>();
+                if (node.getSuperclass() != null) {
+                    builder.add(node.getSuperclass());
+                }
+                return builder
+                    .add(node.getInterfaces())
+                    .build();
+            }
+        }).breadthFirst(sourceObject.getClass());
+
+        // keep only those implemented interfaces which are in model contract set
+        Set<Class<?>> filteredImplementedInterfaces = new HashSet<>();
+        for (Class<?> i : allImplementedInterfaces) {
+            Class<?> actualSubType = potentialModelContractInterfaces.get(i.getName());
+            if (actualSubType != null) {
+                filteredImplementedInterfaces.add(actualSubType);
+            }
+        }
```

Comment:
💭 Do we care to cache the results around this method, in case there are multiple invocations?

Reply:
For now, I've added a shortcut mechanism for skipping most of this work, if there are no contract interfaces specified in the model (which should be most cases). 

We can maybe also add caching in a subsequent PR.

Reply:
@donat wdyt?

Reply:
My initial thought was that we don't need any caching because these computations would be done only once per type. Do we expect running the mapping over and over again? Will the new models expose the same interfaces many times on a single model, or will the client traverse the returned tooling model more than once?

In any case, I think the current implementation is fine and you could deal with caching separately, if you decide to implement it.